File: sensing/preprocessor/gnss/gnss_poser/src/fixposition_converter.cpp
#include <cmath>
#include "ros/ros.h"

#include <Eigen/Geometry>
#include <Eigen/Dense>

#include <std_msgs/Header.h>
#include <nav_msgs/Odometry.h>
#include <geometry_msgs/TwistStamped.h>
#include <geometry_msgs/TwistWithCovarianceStamped.h>
#include <ublox_msgs/NavPVT.h>


static ros::Subscriber odom_sub;
static ros::Publisher twist_pub;
static ros::Publisher twist_cov_pub;
static ros::Publisher pvt_pub;


static double radianToDegree(double radian)
{
    double pi = 3.14159;
    return(radian * (180 / pi));
}

static void odomCallback(const nav_msgs::OdometryConstPtr &msg)
{
    geometry_msgs::TwistStamped current_twist_msg;
    geometry_msgs::TwistWithCovarianceStamped current_twist_convariances_msg;
    ublox_msgs::NavPVT current_navpvt_msg;
    std_msgs::Header header_msg;

    header_msg.frame_id = "base_link";
    header_msg.stamp = msg->header.stamp;
    
    // twist w/o cov
    current_twist_msg.twist = msg->twist.twist;
    current_twist_msg.header = header_msg;
    twist_pub.publish(current_twist_msg);

    // twist w/ cov
    current_twist_convariances_msg.twist = msg->twist;
    current_twist_convariances_msg.header = header_msg;
    twist_cov_pub.publish(current_twist_convariances_msg);

    // pvt
    Eigen::Quaternionf quad_t(msg->pose.pose.orientation.w,
        msg->pose.pose.orientation.x,
        msg->pose.pose.orientation.y,
        msg->pose.pose.orientation.z
        );
    quad_t.normalize();
    auto euler = quad_t.toRotationMatrix().eulerAngles(0, 1, 2);
    current_navpvt_msg.heading =  int(radianToDegree(euler[2])*100000);
    pvt_pub.publish(current_navpvt_msg);

}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "fixposition_converter_node");
    ros::NodeHandle nh;
    odom_sub = nh.subscribe("/fixposition/odometry_enu", 10, odomCallback);
    twist_pub = nh.advertise<geometry_msgs::TwistStamped>("/localization/twist_estimator/twist", 5);
    twist_cov_pub = nh.advertise<geometry_msgs::TwistWithCovarianceStamped>("/localization/twist_estimator/twist_with_covariance", 5);
    pvt_pub = nh.advertise<ublox_msgs::NavPVT>("/fixposition/navpvt", 5);
    ros::spin();
    return 0;
}

File: sensing/preprocessor/gnss/gnss_poser/src/fixposition_converter.py
#!/usr/bin/python
import rospy

import numpy as np

from nav_msgs.msg import Odometry
from geometry_msgs.msg import TwistStamped
from geometry_msgs.msg import Twist
from geometry_msgs.msg import TwistWithCovarianceStamped
from std_msgs.msg import Header
import math

from ublox_msgs.msg import NavPVT

# transform between sensor and vehicle
trans_matrix = np.array([
    [0.999996, 0.002721, 0.001026, 0.287821],
    [0.002728, -0.999969, -0.007332, -0.437961],
    [0.001006, 0.007334, -0.999973, -0.322812],
    [0, 0, 0, 1]])

# rotation_matrix = np.array([
#     [0.999996, 0.002721, 0.001026],
#     [0.002728, -0.999969, -0.007332],
#     [0.001006, 0.007334, -0.999973]])

rotation_matrix = np.array([
    [1.0000000, 0.0000000, 0.0000000],
    [0.0000000, 0.9993908, -0.0348995],
   [0.0000000,  0.0348995,  0.9993908]])

def tranform_twist(rot_mat, untransed_twist):
    transed_twist = Twist()
    untransed_omega = np.array([untransed_twist.angular.x, untransed_twist.angular.y, untransed_twist.angular.z])
    transed_omega = np.matmul(rot_mat.T, untransed_omega.T)
    transed_twist.linear.x = untransed_twist.linear.x
    # transed_twist.angular.x = transed_omega[0]
    # transed_twist.angular.y = transed_omega[1]
    transed_twist.angular.z = transed_omega[2]
    return transed_twist
    


def eulerFromQuaternion(x, y, z, w):
    t0 = +2.0 * (w * x + y * z)
    t1 = +1.0 - 2.0 * (x * x + y * y)
    roll_x = math.atan2(t0, t1)
     
    t2 = +2.0 * (w * y - z * x)
    t2 = +1.0 if t2 > +1.0 else t2
    t2 = -1.0 if t2 < -1.0 else t2
    pitch_y = math.asin(t2)
     
    t3 = +2.0 * (w * z + x * y)
    t4 = +1.0 - 2.0 * (y * y + z * z)
    yaw_z = math.atan2(t3, t4)
     
    return roll_x, pitch_y, yaw_z

def get_quaternion_from_euler(roll, pitch, yaw):
    """
    Convert an Euler angle to a quaternion.
    
    Input
      :param roll: The roll (rotation around x-axis) angle in radians.
      :param pitch: The pitch (rotation around y-axis) angle in radians.
      :param yaw: The yaw (rotation around z-axis) angle in radians.
 
    Output
      :return qx, qy, qz, qw: The orientation in quaternion [x,y,z,w] format
    """
    qx = np.sin(roll/2) * np.cos(pitch/2) * np.cos(yaw/2) - np.cos(roll/2) * np.sin(pitch/2) * np.sin(yaw/2)
    qy = np.cos(roll/2) * np.sin(pitch/2) * np.cos(yaw/2) + np.sin(roll/2) * np.cos(pitch/2) * np.sin(yaw/2)
    qz = np.cos(roll/2) * np.cos(pitch/2) * np.sin(yaw/2) - np.sin(roll/2) * np.sin(pitch/2) * np.cos(yaw/2)
    qw = np.cos(roll/2) * np.cos(pitch/2) * np.cos(yaw/2) + np.sin(roll/2) * np.sin(pitch/2) * np.sin(yaw/2)
 
    return [qx, qy, qz, qw]



class FixpositionConverter:
    def __init__(self):


        self.sub_current_odom = rospy.Subscriber('/fixposition/odometry_enu', Odometry, self.current_odom_callback, queue_size=10)

        self.pub_twist = rospy.Publisher('/fixposition/twist_estimator/twist', TwistStamped, queue_size=2, latch=False)
        self.pub_twist_with_covariance = rospy.Publisher('/fixposition/twist_estimator/twist_with_covariance', TwistWithCovarianceStamped, queue_size=2, latch=False)
        self.pub_navpvt = rospy.Publisher('/fixposition/navpvt', NavPVT, queue_size=2, latch=False)

    def current_odom_callback(self, msg):
        current_twist_convariances_msg = TwistWithCovarianceStamped()
        current_twist_msg = TwistStamped()
        current_navpvt_msg = NavPVT()

        header_msg = Header()
        header_msg.frame_id = "base_link"
        header_msg.stamp = msg.header.stamp

        current_twist_convariances_msg.header = header_msg
        # transed_twist = tranform_twist(rotation_matrix, msg.twist.twist)
        transed_twist = msg.twist.twist
        current_twist_convariances_msg.twist = msg.twist
        current_twist_convariances_msg.twist.twist = transed_twist

        current_twist_msg.header = header_msg
        current_twist_msg.twist = transed_twist

        temp_x = msg.pose.pose.orientation.x
        temp_y = msg.pose.pose.orientation.y
        temp_z = msg.pose.pose.orientation.z
        temp_w = msg.pose.pose.orientation.w
        euler = eulerFromQuaternion(temp_x, temp_y, temp_z, temp_w) 

        current_navpvt_msg.heading = int(math.degrees(euler[2])*100000)
        self.pub_twist.publish(current_twist_msg)
        self.pub_twist_with_covariance.publish(current_twist_convariances_msg)
        self.pub_navpvt.publish(current_navpvt_msg)
    

if __name__ == '__main__':
    rospy.init_node("fixpostion_converter", anonymous=True)
    fixposition_converter_node = FixpositionConverter()
    rospy.spin()
